jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/info/ThreadGroupInfo.java

```diff
@@ -0,0 +1,209 @@
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+package openj9.tools.attach.diagnostics.info;
+
+import java.io.IOException;
+import java.lang.management.LockInfo;
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadInfo;
+import java.lang.management.ThreadMXBean;
+import java.util.Properties;
+
+import com.ibm.tools.attach.target.IPC;
+
+import openj9.tools.attach.diagnostics.base.DiagnosticProperties;
+import openj9.tools.attach.diagnostics.base.DiagnosticsInfo;
+
+/**
+ * Class for encoding and decoding stack traces for transmission via attach API.
+ * This mimics the behaviour of Thread
+ */
+public class ThreadGroupInfo implements DiagnosticsInfo {
+
+	private static final String TIME_FORMAT_STRING = "\t%s time: %d ns%n";
+	static final String THREAD_GROUP_PREFIX = "threadgroup."; //$NON-NLS-1$
+	static final String CPU_TIME = "cpu_time"; //$NON-NLS-1$
+	static final String USER_TIME = "user_time"; //$NON-NLS-1$
+	private static final String THREAD_GROUP_VERSION = THREAD_GROUP_PREFIX + "version"; //$NON-NLS-1$
+	private static final String THREAD_GROUP_JAVA_INFO = THREAD_GROUP_PREFIX + DiagnosticsInfo.JAVA_INFO; //$NON-NLS-1$
+	private static final String NUM_THREADS = THREAD_GROUP_PREFIX+"num_threads"; //$NON-NLS-1$
+	private static ThreadMXBean threadBean;
+	
+	private final int numThreads;
+	private final JvmThreadInfo[] myThreads;
+	private final String groupInfoVersion;
+	private final DiagnosticProperties myProps;
+	private final String javaInfo;
+	private boolean addSynchronizers;
+	
+	/**
+	 * Create an object from received properties
+	 * @param props Properties receive from the target
+	 * @param printSynchronizers add ownable synchronizers to output
+	 * @throws IOException if properties file is corrupt
+	 */
+	public ThreadGroupInfo(Properties props, boolean printSynchronizers) throws  IOException  {
+		addSynchronizers = printSynchronizers;
+		try {
+			DiagnosticProperties.dumpPropertiesIfDebug("Thread group:", props); //$NON-NLS-1$
+			myProps = new DiagnosticProperties(props);
+			if (myProps.containsField(IPC.PROPERTY_DIAGNOSTICS_ERROR) && myProps.getBoolean(IPC.PROPERTY_DIAGNOSTICS_ERROR)) {
+				final String targetMsg = myProps.getProperty(IPC.PROPERTY_DIAGNOSTICS_ERRORMSG);
+				String msg = myProps.getProperty(IPC.PROPERTY_DIAGNOSTICS_ERRORTYPE); 
+				if ((null != targetMsg) && !targetMsg.isEmpty()) {
+					msg = msg+ ": " +  targetMsg; //$NON-NLS-1$
+				}
+				throw new IOException(msg);
+			}
+			numThreads = myProps.getInt(NUM_THREADS);
+			groupInfoVersion = props.getProperty(THREAD_GROUP_VERSION);
```

Comment:
Seems there should be a check of the version, with an appropriate action.

Reply:
Adding a check.